### PR TITLE
Sum and multiply ops with scalar tensors

### PR DESCRIPTION
**Pull request by AlbertMitjans**
Support sums and products of `Operator` classes with scalar tensors of any interface (numpy, jax, tensorflow, torch...).

  ```pycon
  >>> s_prod = torch.tensor(4) * qml.RX(1.23, 0)
  >>> s_prod
  4*(RX(1.23, wires=[0]))
  >>> s_prod.scalar
  tensor(4)
  ```

**Note**: Importing `SUPPORTED_INTERFACES` in `operation.py` caused a circular import, which could be solved in two ways:
- Moving `SUPPORTED_INTERFACES` to the `pennylane/__init__.py` file
- Removing the `Device` and `QuantumTape` types in `execution.py`

I did the second.

### CODE REVIEW

File: pennylane/numpy/tensor.py [hunk 1]
```diff
@@ -154,6 +153,9 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
         # of the vectorized ufunc
         res = super().__array_ufunc__(ufunc, method, *args, **kwargs)
 
+        if isinstance(res, Operator):
```
Comment (albi3ro) [inline pennylane/numpy/tensor.py:156]: I was just thing about this.  Maybe if we don't want this code to be dependent on `Operator` and we want it to be open to extension:
```suggestion
        if not isinstance(res, _np.ndarray):
```

Comment (albi3ro) [reply pennylane/numpy/tensor.py:156]: Ignore that doesn't work ;(

File: pennylane/interfaces/execution.py [hunk 2]
```diff
@@ -32,8 +32,6 @@
 from cachetools import LRUCache
 
 import pennylane as qml
-from pennylane import Device
-from pennylane.tape import QuantumTape
```
Comment (AlbertMitjans) [inline pennylane/interfaces/execution.py:36]: These types caused a circular import.

File: pennylane/operation.py [hunk 3]
```diff
@@ -102,18 +102,19 @@
 from enum import IntEnum
 from typing import List
 
+import autoray
 import numpy as np
 from numpy.linalg import multi_dot
 from scipy.sparse import coo_matrix, eye, kron
 
 import pennylane as qml
+from pennylane.interfaces import SUPPORTED_INTERFACES
```
Comment (AlbertMitjans) [inline pennylane/operation.py:111]: The circular import was originated here. Should we move `SUPPORTED_INTERFACES` to the `pennylane/__init__.py` file instead?

**Review by albi3ro — COMMENTED** [review]
So it seems like the circular import is coming from a reliance on `pennylane.interfaces.SUPPORTED_INTERFACES`.

My question: does `operation.py` *need* to use `SUPPORTED_INTERFACES`? 

Maybe we are being too restrictive with validation.

For example, we want to start having more vectorization (parameter batching).  In that case, we will want to allow things like:
```
torch.tensor([1.2, 2.3]) * qml.PauliX(0)
```
So we don't need the check on shape.

I vote we leave it up to `SProd` to validate what it allows for scalar coefficients.

**Comment by AlbertMitjans** [comment]
> I vote we leave it up to SProd to validate what it allows for scalar coefficients.

I like this idea! Moved the check into the `SProd` class.

> So we don't need the check on shape.

I am currently checking the shape, but I added a `TODO` message saying to remove this check when batching is supported.

**Review by albi3ro — APPROVED** [review]
Looks good 👍

**Pull request merged**